client-tools/Dump/DumpFeature.cpp

```diff
@@ -1364,4 +1470,349 @@ void DumpFeature::start() {
+      "writer threads (+) / network threads (-)",
+      "dbserver worker put batch (+) / rest handler get batch (-)",
+  };
+
+  std::string msg;
+  for (size_t i = 0; i < 2; i++) {
+    if (i > 0) {
+      msg += ", ";
+    }
+    msg += locations[i];
+    msg += " = ";
+    msg += std::to_string(blockCounter[i]);
+  }
+
+  LOG_TOPIC("d1349", INFO, Logger::DUMP) << "block counter " << msg;
+}
+
+void DumpFeature::ParallelDumpServer::ParallelDumpServer::countBlocker(
+    BlockAt where, int64_t c) {
+  /* clang-format off */
+  const char* locations[] = {
+      "writer threads - consider increasing the number of network threads",
+      "network threads - consider increasing the number of local writer threads",
+      "dbserver get batch - consider increasing the parallelism on dbservers",
+      "dbserver put batch - consider increasing the number of network threads",
+  };
+  /* clang-format on */
+  const char* msg = nullptr;
+  auto actual = blockCounter[where].fetch_add(c);
+  if (actual == 100) {
+    msg = locations[2 * where];
+    blockCounter[where].fetch_sub(100);
+  } else if (actual == -100) {
+    msg = locations[2 * where + 1];
+    blockCounter[where].fetch_add(100);
+  }
+
+  if (msg) {
+    LOG_TOPIC("3cc53", DEBUG, Logger::DUMP)
+        << "when dumping data from server " << server << " system blocking at "
+        << msg;
+  }
+}
+
+DumpFeature::ParallelDumpServer::ParallelDumpServer(
+    ManagedDirectory& directory, DumpFeature& feature,
+    ClientManager& clientManager, const DumpFeature::Options& options,
+    maskings::Maskings* maskings, DumpFeature::Stats& stats,
+    std::shared_ptr<DumpFileProvider> fileProvider,
+    std::unordered_map<std::string, ShardInfo> shards, std::string server)
+    : DumpJob(directory, feature, options, maskings, stats,
+              VPackSlice::noneSlice()),
+      clientManager(clientManager),
+      fileProvider(std::move(fileProvider)),
+      shards(std::move(shards)),
+      server(std::move(server)),
+      queue(options.localWriterThreads) {}
+
+std::unique_ptr<httpclient::SimpleHttpResult>
+DumpFeature::ParallelDumpServer::receiveNextBatch(
+    httpclient::SimpleHttpClient& client, std::uint64_t batchId,
+    std::optional<std::uint64_t> lastBatch) {
+  std::string url = basics::StringUtils::concatT(
+      "_api/dump/next/", dumpId, "?batchId=", batchId, "&dbserver=", server);
+  if (lastBatch) {
+    url += "&lastBatch=" + std::to_string(*lastBatch);
+  }
+
+  while (true) {
+    std::unique_ptr<arangodb::httpclient::SimpleHttpResult> response(
+        client.request(arangodb::rest::RequestType::POST, url, nullptr, 0, {}));
+    auto check = ::arangodb::HttpResponseChecker::check(
+        client.getErrorMessage(), response.get());
+    if (check.fail()) {
+      LOG_TOPIC("ad972", ERR, arangodb::Logger::DUMP)
+          << "An error occurred while dumping from server '" << server
+          << "': " << check.errorMessage();
+
+      // TODO complete this list!
```

Comment:
How about the error that happens when a connection is broken? This sounds important to me in the cloud business...

Reply:
All others are trivial. This one should at least handle the connection broken case, in particular since it is simple to implement, because the retry logic is already there.

Reply:
Sadly the SimpleHttpClient has a very intransparent error handling. Many errors result in a `DEAD` connection, i.e. the response is `nullptr`. Connection refused, connection broken or DNS resolution failure. Fixing all these things would require more work.

Reply:
OK, what a sad state of affairs!